### DISCUSSION TITLE
umi4使用monorepoRedirect的问题

### DISCUSSION THREAD

**Discussion by buquan** [umijs/umi#10892] [General]
请教：
1.app项目使用monorepoRedirect: {} 但loader编译不过(包入口源文件入口是src/index.ts,但某一个子文件夹文件是jsx),还需要配置extraBabelIncludes: ['包名'] 才可以是为什么呢
2. 即使使用monorepoRedirect: {}，但app项目build时候package.json里还是需要声明'包名': workspace:*吗
3. monorepoRedirect这个srcDir意思是说指定的默认src如果找不到,会去包的packagejson的字段找嘛

**Comment by fz6m** [comment]
1. `某一个子文件夹文件是jsx` 编译不过，需要给一个最小复现。

2. 在现在的设计里，不对 `"deps": "workspace:*"` 有强要求，因为包管理工具不一样，有的不支持 workspace 协议，所以当你在 `package.json` 里声明了仓库内其他的子包，无论他的版本是多少，都会重定向（可以通过 `monorepoRedirect.exclude` 选项排除）。

3. 不会去 `package.json` 找，也不需要修改子包 `package.json` ，和 `package.json` 没有关系。正常来说你的子包产物应该在 `"main": "dist/index.js"` ，为了调试方便 + 热更新，所以 `monorepoRedirect` 帮你把包的 alias 改到了 `src/index` 下面，所以你不需要修改你子包的 `package.json` 。

**Reply by buquan** [reply]
感谢回复，对于3，还有个疑问，如果源码不是src/index,会alias到哪

**Reply by fz6m** [reply]
默认是 `alias` 到子包的 `src` 文件夹，入口可以是 `src/index.{js,jsx,ts,jsx}` ，如果你不在 `src` 文件夹写源码，可以改 `srcDir: ['lib']` （此时重定向到 `lib` ）

**Reply by buquan** [reply]
大概了解了 感谢解惑

**Reply by buquan** [reply]
<img width="1204" alt="image" src="https://user-images.githubusercontent.com/23711051/229488664-490be49c-4a7c-44b3-846a-f5cd27f8a5d8.png">

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/23711051/229487383-9e1ca7ce-8fc8-4965-bbf5-c493bd39b380.png">


@fz6m 对于`某一个子文件夹文件是jsx 编译不过，需要给一个最小复现`，麻烦看下ui的AuthDevice有什么问题, 目前情况是只有开启`extraBabelIncludes`才可以正常编译，否则就会报图里的错


[ui的lib库源码](https://github.com/buquan/ifss/tree/master/packages/common-ui)

**Reply by fz6m** [reply]
1. `MyButton/index.tsx` ：开启了 `monorepoRedirect` 之后，第三方包会和项目变成一体，此时的 `@` 是项目本身的 `src` 文件夹，不再是子包的别名，此时子包的别名是不知道的，这个是 `monorepoRedirect` 已知的限制，这意味着子包都需要使用相对路径导入，否则请使用 `father dev` 这样的变化监听的方式实时生成最新的产物，无需开启 `monorepoRedirect` 。

2. `MyDesc/index.js` ：所有的节点必须写在 `x` 文件里，比如 `jsx/tsx` ，只有 `x` 结尾的才可以写节点标签。

**Reply by buquan** [reply]
@fz6m 非常感谢指点，

按照你说的当`monorepoRedirect `监听`src`时，lib组件库的组件如果是标签需要改成带`x`的，还有把`@`改了相对路径后跑起来了，而且自测了umi4不能监听`node_modules`热更新,打印到`exclude: [/node_modules/]`了。

-> umi3对于`js jsx`都可以编译`config.module.rule('js') test: /\.(js|mjs|jsx|ts|tsx)$/,`，
-> umi4对于monorepo`config.module.rule('jsx-ts-tsx') test: /\.(jsx|ts|tsx)$/` 没有针对js的标签编译，之前加了`extraBabelIncludes`可以用是因为`config.module.rule('extra-src') test: /\.(js|mjs|cjs)$/`有js等于是这俩一起把所有文件编译才跑过

[复现的最小仓库](https://github.com/buquan/ifss/) `pnpm i ` `pnpm:start:umi4app`

不知道理解的对不对，而且有个困惑就是monorepoRedirect不编译js是有什么特殊的原因吗

**Reply by fz6m** [reply]
不是不编译 js ，是因为用子包的时候实际上是在 `node_modules` 路径下面，这下面的 js 文件会被视为产物略过，所以才会有 `monorepoRedirect` 快速解决 monorepo 的子包热更新问题。
